frame/pablo/src/lib.rs

```diff
@@ -467,6 +515,83 @@ pub mod pallet {
 		pub(crate) fn account_id(pool_id: &T::PoolId) -> T::AccountId {
 			T::PalletId::get().into_sub_account(pool_id)
 		}
+		pub(crate) fn do_get_exchange_rate(
+			pool_id: T::PoolId,
+			swap: bool,
+		) -> Result<Rate, DispatchError> {
+			let pair = Self::currency_pair(pool_id)?;
+			let pool_account = Self::account_id(&pool_id);
+			let pair = if swap { pair.swap() } else { pair };
+			let pool_base_asset_under_management =
+				T::Convert::convert(T::Assets::balance(pair.base, &pool_account));
+			let pool_quote_asset_under_management =
+				T::Convert::convert(T::Assets::balance(pair.quote, &pool_account));
+			Ok(Rate::checked_from_rational(
+				pool_base_asset_under_management,
+				pool_quote_asset_under_management,
+			)
+			.ok_or(ArithmeticError::Overflow)?)
+		}
+
+		fn update_twap(pool_id: T::PoolId) -> Result<(), DispatchError> {
+			if TWAPState::<T>::contains_key(pool_id) {
+				let current_timestamp = T::Time::now();
+				let rate_base = Self::do_get_exchange_rate(pool_id, false)?;
+				let rate_quote = Self::do_get_exchange_rate(pool_id, true)?;
+				TWAPState::<T>::get(pool_id).map_or(
+					Ok(()),
+					|previous_cumulative_price: TWAPStateOf<T>| -> Result<(), DispatchError> {
+						let ((elapsed, base_price_cumulative), (_, quote_price_cumulative)) = (
+							compute_next_price_cumulative::<T::Convert, _, _>(
+								previous_cumulative_price.timestamp,
+								previous_cumulative_price.base_price_cumulative,
+								current_timestamp,
+								rate_base,
+							)?,
+							compute_next_price_cumulative::<T::Convert, _, _>(
+								previous_cumulative_price.timestamp,
+								previous_cumulative_price.quote_price_cumulative,
+								current_timestamp,
+								rate_quote,
+							)?,
+						);
```

Comment:
Why doing this computation here if we check that elapsed >= TWAPInterval after? Doesn't this mean we are going to waste this sometimes?

Reply:
Alright I think what we are doing here is somehow wrong, the cumulative should always be updated, only the actual price computation should be done when the `interval` passed

Reply:
Good catch, in fact we are not calculating an average price here because according to the formula and applying the current code,
`twap = [(Price_end * elapsed + previous_cumulative) - previous_cumulative ] / elapsed = Price_end`

Therefore every time spot price changes(exchange and possibly addLiqudity, removeLiquidity) the cumulative should be calculated while keeping other fields of `TimeWeightedAveragePrice` struct in tact. We need to write some good tests here which changes prices during the configured interval, which affects the TWAP.